python/mxnet/contrib/amp/amp.py

```diff
@@ -0,0 +1,342 @@
+__all__ = ['init', 'init_trainer', 'scale_loss', 'unscale']
+
+from types import MethodType
+import logging
+import contextlib
+import numpy as np
+
+from ... import symbol
+from ...symbol import Symbol
+from ...symbol import contrib as symbol_contrib
+from ... import ndarray
+from ...ndarray import NDArray
+from . import lists
+from ...gluon import trainer
+from ... import base
+from ... import optimizer as opt
+from .loss_scaler import LossScaler
+
+def _cast_symbol_NDArray(s, dtype):
+    float_types = (np.float16, np.float32)
+    if isinstance(s, Symbol):
+        return symbol.amp_cast(s, dtype=dtype)
+    elif isinstance(s, NDArray):
+        if (s.dtype != dtype and
+                s.dtype in float_types and
+                s.context.device_type != 'cpu'):
+            return ndarray.amp_cast(s, dtype=dtype)
+        else:
+            return s
+    else:
+        return s
+
+def _get_fun_to_wrap(name, module, submodule_dict):
+    module_internal = getattr(module, "_internal")
+    prefix = base._get_op_name_prefix(name)
+    if len(prefix) > 0:
+        if prefix != '_random_' or name.endswith('_like'):
+            func_name = name[len(prefix):]
+            cur_module = submodule_dict[prefix]
+        else:
+            func_name = name
+            cur_module = module_internal
+    elif name.startswith('_'):
+        func_name = name
+        cur_module = module_internal
+    else:
+        func_name = name
+        cur_module = module
+    return func_name, cur_module
+
+def _wrap_symbol_functions(module, target_dtype, target_precision_ops=None,
+                           conditional_fp32_ops=None, fp32_ops=None):
+    def _ndarray_wrapper(f, target_dtype, cond_arg=None):
+        def _new_fun(*args, **kwargs):
+            if cond_arg is not None:
+                if (cond_arg[0] not in kwargs or
+                        kwargs[cond_arg[0]] not in cond_arg[1]):
+                    return f(*args, **kwargs)
+            new_args = list(map(lambda x: _cast_symbol_NDArray(x, target_dtype), args))
+            args = tuple(new_args)
+            kwargs = {k: _cast_symbol_NDArray(v, target_dtype) for k, v in kwargs.items()}
+            return f(*args, **kwargs)
+        _new_fun.__name__ = f.__name__
+        _new_fun.__module__ = f.__module__
+        _new_fun.__doc__ = f.__doc__
+        return _new_fun
+
+    def _symbol_wrapper(f, target_dtype, cond_arg=None):
+        def _new_fun(*args, **kwargs):
+            if cond_arg is not None:
+                if (cond_arg[0] not in kwargs or
+                        kwargs[cond_arg[0]] not in cond_arg[1]):
+                    return f(*args, **kwargs)
+            sym = f(*args, **kwargs)
+            inputs = sym.get_children()
+            aux = sym.list_auxiliary_states()
+            inputs = list(map(lambda x: _cast_symbol_NDArray(x, target_dtype)
+                              if x.name not in aux else x, inputs))
+            atomic_sym = sym._gen_atomic_symbol()
```

Comment:
why is this needed ?

Reply:
Basically this allows us to cast inputs to the op that were not specified by the user (e.g. you make a convolution via symbolic API, you don't need to pass weights and biases to it, MXNet will generate them). So what we do here is create a symbol using the original function (which will create all the other children), then take those children and cast them. MXNet (nor NNVM) does not allow, however, to manipulate a symbol children after it was created. So what we do is we create a new, atomic symbol (atomic means it does not have any children set yet, but has all the params the same as the symbol used to create it), and populate that symbol with casted inputs.

Reply:
when pretrained models are loaded under amp.init will it silently fail ?

Reply:
Pretrained as in symbol? That will not do anything (it needs your PR).
Pretrained as in load_parameters? That will work as expected (I was using it for all my experiments with GluonCV where e.g. SSD uses pretrained RN50 backbone).